### ISSUE TITLE
No date in form in duration mode

### ISSUE THREAD

**Issue by SaturnusDJ** [kimai/kimai#3993]
### Describe the issue

Checked in Kimai demo.

When time tracking mode is set to duration, there is no date field when adding time via 'My times'. 

![image](https://user-images.githubusercontent.com/14320018/235078727-b18e48a0-a4ff-4a75-aa11-873e431418eb.png)

In the documentation at https://www.kimai.org/documentation/configurations.html it is suggested that there should be a date field.

>Duration (no end time)
>
>Removes the end fields from the timesheet form, while the duration can be edited. All timesheet >tables will only display the date and duration for all records, start and end time columns are hidden.
>
With default mode as premise, when the end fields are removed, the date field should still be there. (That's why this is set up as bug report too.)

I've looked in issues and discussions for information about bugs, changes and design choices, but none found.

Practically speaking... In a company here, the exact times do not matter, however it does happen that an employee has to create a record for a different day than today. That is not possible now.

### I already tried

- [X] I've read and searched [the documentation](https://www.kimai.org/documentation/).
- [X] I've searched for similar issues in this repository.
- [X] I've searched for similar issues in [the discussions](https://github.com/kimai/kimai/discussions).

### Kimai version

2.0.15

### How do you run Kimai?

Docker

### Which PHP version are you using?

8.1

**Comment by kevinpapst** [comment]
If you allow to edit the date and duration, then you directly use the default mode.

But thanks for pointing out the outdated docs 👍 
Those docs were written for 1.x and I updated them now to match 2.x.

**Comment by SaturnusDJ** [comment]
CSS is used now to hide the start and end times. This does come with strong downsides:
1. When someone would make a long working day, say 07:00-17:00, and enters this data at 17:00, it will be entered as 17:00+10 hours, resulting in hours booked across 2 days. This is very wrong.
2. There will be many start and end times without meaning. Everything being put to 0:00 makes more sense (unity).

Basically, duration mode is the way to go here, but plus the ability to choose the date. So that users who did not enter their hours for a certain day yet (forgot, busy, working remotely, emergency, etc.), can still do that another day.

**Comment by kevinpapst** [comment]
I would recommend to configure the default start time under System > Settings to a fixed time, like 10:00.
That solves 1 and 2 and is mentioned in the docs.

**Comment by SaturnusDJ** [comment]
>Default start-time (not used in all timetracking modes)

This does not work for timetracking mode 'default', which is what we **have** to use in order to allow people to change the date.

**Comment by kevinpapst** [comment]
Ah, right. You should be able to use a custom JS to set the time.

**Comment by SaturnusDJ** [comment]
```
var button = document.getElementsByClassName("btn btn-white action-create modal-ajax-form");
button[0].addEventListener("click",function(e){
    document.getElementById("timesheet_edit_form_begin_time").value="00:12";
},false);
```
Is as far as I could come in >3 hours as programming noob.

https://stackoverflow.com/questions/17633152/handling-onclick-event-with-pure-javascript
https://stackoverflow.com/questions/4683331/how-to-fill-in-form-field-and-submit-using-javascript

line 1 sets the var button to the actual button(s) by the given classname
line 2 selects the first one in the array (also only one because pretty unique classname) and watches for a click event on it
line 3 executed when the click on the button is done: set the value of the field
line 4 closings

Detecting the click on the button works, but setting the field does not. However when executed by itself (paste line 3 in Firefox - Inspector - Console) it does work. But line 3 cannot work by itself because when https://demo-empty.kimai.org/en_GB/timesheet/ is loaded, the fields of the form are not loaded yet. Only after clicking on the create button, therefore some kind of `eventlistener` is needed I thought. But something happens that makes line 3 result in `null`. 
```
Uncaught TypeError: document.getElementById(...) is null
    <anonymous> https://demo-empty.kimai.org/en_GB/timesheet/:750
    EventListener.handleEvent* https://demo-empty.kimai.org/en_GB/timesheet/:750
[timesheet:750:236](https://demo-empty.kimai.org/en_GB/timesheet/)
    <anonymous> https://demo-empty.kimai.org/en_GB/timesheet/:750
    (Async: EventListener.handleEvent)
    <anonymous> https://demo-empty.kimai.org/en_GB/timesheet/:750
```
Maybe it still is not loaded because it comes from the current time. I could not get a sleep/delay function working. Might take it to StackOverflow.

**Comment by SaturnusDJ** [comment]
Back to the root though, I feel like this (title) should at least be an option. https://github.com/kimai/kimai/issues/3993#issuecomment-1529441984 is not fixed. It makes so much more sense to me to run an organisation where duration can be booked on any date, then to run an organisation where duration can only be entered for today (reasons in same comment). In duration mode, it is even still allowed via 'Weekly hours' to enter for other dates. I see no sense in not being able to choose the date on the form in 'My times'.

**Comment by SaturnusDJ** [comment]
Alternative solution: do use the `Default start-time` in `Timetracking mode` Default. Design wise the easiest and it makes a lot of sense as the setting is already there, but not being used.

**Comment by kevinpapst** [comment]
Hm, my first reaction was: absolutely not. That makes no sense. 
But then, in theory nothing should change for the "Default mode" users... unless someone configured a "static time" but expects "now". What will definitely happen and increase my support time.

**Comment by SaturnusDJ** [comment]
To make it even stranger...
The set time, in `Default start-time` when using `Timetracking mode` Default:
- Is not used in `My times`.
- Is used in `Weekly hours`.
- Is used in `Calendar.`

That is a discrepancy. The description `(not used in all timetracking modes)` refers to the setting above it, not the way of how time entries are made. But even when that would be changed, it stays a discrepancy.

**Comment by kevinpapst** [comment]
Agree. That is strange. Calendar was completely re-written for 2.0. Any yes, the software has bugs. That is normal, every software has bugs. Especially when its for free and no-one wants to pay for maintenance...

Thanks for giving back with your detailed input, time and research!

**Comment by SaturnusDJ** [comment]
No offence taken from the bugs haha. I am the kind of person running into them more often than others. Glad I can be of help with the reports.

What is your direction on this one?

**Comment by kevinpapst** [comment]
This should mainly do what you want. Set the time to a static value and hide the start time:
```javascript
document.addEventListener('show.bs.modal', (e) => { 
    if (e.srcElement.querySelector('form[name=timesheet_edit_form]') === null) { return; }
    const time = e.srcElement.querySelector('#timesheet_edit_form_begin_time'); 
    if (time !== null) { time.value = '01:00'; time.parentElement.style.display = 'none'; } 
});
````

**Comment by SaturnusDJ** [comment]
Thanks for your efforts!! When having set a fixed time in `Default start-time (not used in all timetracking modes)` this works. When _now_ is set it does not work. Just saying this for clarity for whomever reads this topic.

This was tested on [demo-empty.kimai.org](https://demo-empty.kimai.org). I see the plugin _Custom content_ is now "Compatible with Kimai 2.0". This new version I do not see on Gumroad yet. How to get it?

([demo-branch.kimai.org](https://demo-branch.kimai.org) is down by the way atm.)

**Comment by kevinpapst** [comment]
> How to get it?

If you purchased it before: wait. Emails and news will follow.

**Comment by simone-gasparini** [comment]
Hi @kevinpapst, I write in this issue because I've encountered this problem too but with a different behaviour.
In my mind, in the duration mode the date should be visible or not visible based on the "Allow time entries in the future" flag.
At the moment, in the duration mode, what this flag do?

Thanks in advance

**Comment by kevinpapst** [comment]
@simone-gasparini I'd assume it does nothing, because you cannot create future records.

**Comment by simone-gasparini** [comment]
In our case the duration mode is perfect for the daily routine, but we need that our users put vacation days in advance usign the calendar.
Can you consider the use on the "Allow time entries in the future" flag to show or hide only the date field?
I've seen that even in the duration mode, the /create api is called with the date from the day pressed in the calendar.

**Comment by kevinpapst** [comment]
You are looking for a solution for a problem that does NOT arise from the functionality you are talking about.

For holidays there is a much better solution on the horizon.

**Comment by simone-gasparini** [comment]
Ok perfect, we will wait for it.
At this point the functionality of the calendar in the duration mode is messy, because if i click on any day other than today, the create popup appear, I compile all the information with only the desired duration, but the entry goes on today.

**Comment by kevinpapst** [comment]
Hide the calendar (e.g. via custom css), it doesn't make sense in duration mode. 

Or just leave the normal mode active, if you want to use it anyway.

**Comment by stereo9** [comment]
my boys in the office are also pointing this issue out.
i think for most people it is most intuitive to think "i started at x and worked for y hours" rather than start- and endtime based
i always use the weekly sheet, but as my people point out, the behaviour is new in kimai2 and it worked fine in kimai1


1) configurable starting time makes a lot of sense especially when working with the calendar. set starting time and put duration, this is the standard way the calendar was used in our office so far

2) setting to default mode has one important problem, now you can switch three variables, when the start time is changed, the **duration changes automatically** rather than the end time.

i think that we have to use this mode now creates a lot of room for human error that wasnt there before

EDIT: note that this was pointed out to me by two people individually (my team is 15 people) only 1 day after we upgraded, i think its a usability warning sign

**Comment by kevinpapst** [comment]
@simone-gasparini that is the solution to your holiday issue: https://www.kimai.org/store/controlling.html

> the behaviour is new in kimai2 and it worked fine in kimai1

@stereo9 what exactly ? I can't recall changing calculations in the edit window. I am sure your post contains valid points, but it reads like a mixture of 5 different issues, none of which I do unterstand entirely.

**Comment by SaturnusDJ** [comment]
> This should mainly do what you want. Set the time to a static value and hide the start time:
> 
> ```js
> document.addEventListener('show.bs.modal', (e) => { 
>     if (e.srcElement.querySelector('form[name=timesheet_edit_form]') === null) { return; }
>     const time = e.srcElement.querySelector('#timesheet_edit_form_begin_time'); 
>     if (time !== null) { time.value = '01:00'; time.parentElement.style.display = 'none'; } 
> });
> ```

A recent update of Kimai (probably [2.11.0](https://github.com/kimai/kimai/releases/tag/2.11.0)) had a change regarding locales. There is a user preference for this called "Format for time, date and currency".

The following happened: A user changed their password and on saving this change, the locale was set to English, which was already shown in the preferences as the current locale, but this was not actually the current locale. Only upon saving the profile, this was added to the database (verified). The result was a change from the 24 hours system to AM/PM, 'breaking' the above JS-code because e.g. 01:00 is not a valid English time (AM/PM is missing). The solution, of course, was to apply a 24 hour locale. This was done by changing the locale in the database and batch adding it for anyone that had not set it yet. (To not annoy everyone with a request.)

I think in that update it might have been good if the locale was derived from the language preference, if the user had set this, and certainly not show a value that was not active, let alone to apply it on saving. But I don't know how everything works, maybe there was no other way, so this is just open feedback. The impact was low and it was easy to solve.

**Comment by kevinpapst** [comment]
@SaturnusDJ the locale should have been received from the language that was set before. Not though the migration, but through the code when opening the edit profile/preferences screen. 
But lets stay on topic, as I already do not understand what this issue is about and more unrelated answers make it even worse 😁

**Comment by SaturnusDJ** [comment]
This user was on NL language, but got the EN locale upon changing password. Unexpected outcome at least.

Yeah it's offtopic. Deemed it too small for a new issue and with the functional impact only on this JS-code I dropped it here. Next time a new topic.
